### PR TITLE
Implement proof of possesion

### LINKED ISSUE #9: ProvePossession -- challenge index generation
1. Determine the correct randomness to draw from based on challenge period and latest proof of possession
2. Sample randomness from evm GetRandDAO opcode 
3. Use randomness to get challenge index

### LINKED ISSUE #10: ProvePossession -- verify proof
1. Use challenge index with SumTree search (#8 ) to find root to prove against
2. Verify merkle proof argument against root id hash 

Things adding uncertainty
* we should find and import a good merkle proving library which has handled lots of proving.  @nicola has done some investigation into this IIRC.
* we need to figure out cids vs no cids in our proofs

### PR DESCRIPTION
**Pull request by anorth**
Implemented PDPService.provePosession method, with tests.

At present, uses the challenge epoch itself as the random seed rather than fetching chain randomness from that epoch (see #44).

I moved the Merkle tree and proof building code alongside the verification code since now its needed from multiple test files. The compiler should strip un-called functions from these libraries when deploying.

Closes #9. Closes #10.

### CODE REVIEW

File: contracts/src/Cids.sol
Comment (anorth) [inline contracts/src/Cids.sol:19]: @ZenGround0 I tried assembly after your suggestion in the previous PR, but eventually tracked proof failures down to this. I would rather leave it to an expert. I can delete the incorrect assembly if you wish, though it might provide a head start to someone coming in.

File: contracts/src/PDPService.sol
Comment (anorth) [inline contracts/src/PDPService.sol:316]: This is an easy hash to use here, but we could use sha256 if preferred.

File: contracts/src/Proofs.sol
Comment (anorth) [inline contracts/src/Proofs.sol:150]: No changes here, just moved the code.

File: contracts/src/Cids.sol
Comment (ZenGround0) [reply contracts/src/Cids.sol:19]: Its way better to keep it commented out, thank you

File: contracts/src/PDPService.sol
Comment (ZenGround0) [reply contracts/src/PDPService.sol:316]: sha3 lgtm

File: contracts/src/PDPService.sol
Comment (ZenGround0) [inline contracts/src/PDPService.sol:319]: We should be using `findRootIds` here to make use of batching cost reduction.

File: contracts/test/PDPService.t.sol
Comment (ZenGround0) [inline contracts/test/PDPService.t.sol:364]: The protocol constant for max leaves is 1 << 50 so this comment is confusing

File: contracts/test/PDPService.t.sol
Comment (ZenGround0) [inline contracts/test/PDPService.t.sol:419]: It doesn't look like this is the bug causing failure but I think this <= is off by one and should be <.  If we got unlucky and we landed with a challenge offset == leafCount[i], i.e. a leaf count of 5 and a challenge offset of 5 then we'd try to access leafs[5] which would overrun the leaf buffer.

File: contracts/test/PDPService.t.sol
Comment (ZenGround0) [inline contracts/test/PDPService.t.sol:421]:
```suggestion
                    treeOffset = challengeOffset;
                    break;
```

**Review by ZenGround0 — COMMENTED** [review]
This should fix it

File: contracts/src/PDPService.sol
Comment (ZenGround0) [inline contracts/src/PDPService.sol:327]: You can remove this as the record keeper PR should reflect these requirements.  I hadn't planned on storing the actual proof but maybe we should at least add it into the recordkeeper call for the consumer to potentially keep around.

File: contracts/test/ProofUtil.sol
Comment (ZenGround0) [inline contracts/test/ProofUtil.sol:10]: Were you planning to move more stuff here?  If not I don't understand extracting this piece out and leaving the others, I think at least buildProof and buildTree belong here.

**Review by ZenGround0 — COMMENTED** [review]
I'll hold off on approval until you take it out of draft and add more tests but with bug fixes this is LingGTM

File: contracts/test/PDPService.t.sol
Comment (anorth) [reply contracts/test/PDPService.t.sol:364]: I don't want to build a 2^50 merkle tree in a test. I'll clarify this inline constant.

File: contracts/src/PDPService.sol
Comment (anorth) [reply contracts/src/PDPService.sol:319]: It doesn't look to me like there is any efficiency gain from the batched method. To call it, I'll have to loop twice here and accumulate the roots in memory between those iterations.

File: contracts/src/PDPService.sol
Comment (anorth) [reply contracts/src/PDPService.sol:327]: I just meant record that the proof happened.

File: contracts/test/ProofUtil.sol
Comment (anorth) [reply contracts/test/ProofUtil.sol:10]: The move was strictly based on generateLeaves being shared between multiple test files.

I do see value in moving some other small helpers out here too. `buildProofs` (plural) is tricker to home. It's the spec for how proofs should be built off chain, but specific to the PDP service (rather than generic Merkle tree ops that I moved to Proofs.sol).

File: contracts/src/Cids.sol
Comment (anorth) [reply contracts/src/Cids.sol:46]: Added a roundtrip (which fails hard with the bad assembly)

File: contracts/src/Proofs.sol
Comment (ZenGround0) [inline contracts/src/Proofs.sol:153]:
```suggestion
    // The last array is the leaves, and each prior array is the result of the hash of pairs in the previous array.
```

**Review by ZenGround0 — APPROVED** [review]

File: contracts/test/PDPService.t.sol
Comment (ZenGround0) [inline contracts/test/PDPService.t.sol:417]: I'm not sure about this but my best understanding is the co prime condition needs to exist over leaf count1 and (leaf count1 + leaf count2)

**Pull request merged**
